src/main/java/org/jboss/ejb/protocol/remote/RemotingEJBDiscoveryProvider.java

```diff
@@ -592,7 +592,18 @@ public void addMatch(final ServiceURL serviceURL) {
         public void cancel() {
             final List<Runnable> cancellers = this.cancellers;
             synchronized (cancellers) {
-                for (Runnable canceller : cancellers) {
+                /**
+                 In scenario in which the last node performs a cancel operation and discovery fails a retry is
+                 attempted (see phase2 in countdown method above). This triggers opening new connections and,
+                 as a result, adding cancellers, which are being iterated over in this method leading to
+                 ConcurrentModificationException. To avoid this the copy is created.
+
+                 See https://issues.redhat.com/browse/JBEAP-24568
+                 */
+                final List<Runnable> cancellersCopy = new ArrayList<>();
+                cancellersCopy.addAll(cancellers);
```

Comment:
The above 2 lines can be combined to 1:

final List<Runnable> cancellersCopy = new ArrayList<>(cancellers);

Reply:
fixed